xds/internal/xdsclient/xdsresource/name.go

```diff
@@ -0,0 +1,115 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xdsresource
+
+import (
+	"net/url"
+	"sort"
+	"strings"
+
+	"google.golang.org/grpc/internal/envconfig"
+)
+
+// Name contains the parsed component of an xDS resource name.
+type Name struct {
+	Scheme    string
+	Authority string
+	Type      string
+	ID        string
+
+	ContextParams map[string]string
+}
+
+// ParseName splits the name and returns a struct representation of the Name.
+//
+// If the name isn't a valid new-style xDS name, field ID is set to the input.
```

Comment:
Can you give examples of what the input might be and how it maps to the output?

Is it impossible to have an invalid input?  Or do we just return `Name.ID=input` and have the caller figure out whether there is an error?

Reply:
Example is in the struct godoc.

I'm leaving it to the caller to figure out if the input is valid.
For now, all invalid input returns `Scheme=""` (note that it is still a valid old-style non-federation resource name)